nix/ogmios-datum-cache-nixos-module.nix

```diff
@@ -0,0 +1,208 @@
+    };
+
+    ogmiosAddress = mkOption {
+      description = "Host address where Ogmios is listening.";
+      type = str;
+      default = "127.0.0.1";
+    };
+
+    ogmiosPort = mkOption {
+      description = "Port where Ogmios is listening.";
+      type = port;
+      default = 1337;
+    };
+
+    blockFilter = mkOption {
+      description = "Filter.";
+      type = nullOr str;
+      default = null;
+    };
+
+    blockSlot = mkOption {
+      description = "Slot of first block to fetch by initial block fetcher.";
+      type = nullOr int;
+      default = null;
+    };
+
+    blockHash = mkOption {
+      description = "Hash of block's HEADER (not hash of a block itself).";
+      type = nullOr str;
+      default = null;
+    };
+
+    fromOrigin = mkOption {
+      description = "Start block fetcher from origin.";
+      type = bool;
+      default = false;
+    };
+
+    fromTip = mkOption {
+      description = "Start block fetcher from chain tip.";
+      type = bool;
+      default = false;
+    };
+
+    extraArgs = mkOption {
+      description = "Extra arguments to ogmios command.";
+      type = listOf str;
+      default = [ ];
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [{
+      assertion = cfg.fromOrigin || cfg.fromTip
+        || (cfg.blockSlot != null && cfg.blockHash != null);
+      message =
+        "Either fromOrigin or fromTip or both blockSlot and blockHash options need to be set.";
+    }];
+
+    services.ogmios-datum-cache.dbConnection =
+      mkIf config.services.postgresql.enable (mkDefault
+        "host=/run/postgresql port=${
+          toString config.services.postgresql.port
+        } dbname=${cfg.dbName}");
+
+    users.users.ogmios-datum-cache = mkIf (cfg.user == "ogmios-datum-cache") {
+      isSystemUser = true;
+      group = cfg.group;
+    };
+    users.groups.ogmios-datum-cache =
+      mkIf (cfg.group == "ogmios-datum-cache") { };
+
+    services.postgresql = mkIf cfg.configurePostgresql {
+      ensureDatabases = [ cfg.dbName ];
+      ensureUsers = [{
+        name = cfg.user;
+        ensurePermissions = { "DATABASE '${cfg.dbName}'" = "ALL PRIVILEGES"; };
+      }];
+    };
+
+    systemd.services.ogmios-datum-cache = {
+      enable = true;
+
+      after = [ "postgresql.service" "ogmios.service" ];
+      wantedBy = [ "multi-user.target" ];
+
+      script = escapeShellArgs (concatLists [
+        [ "${cfg.package}/bin/ogmios-datum-cache" ]
+        [ "--log-level" cfg.logLevel ]
+        [ "--server-port" cfg.port ]
+        [ "--ogmios-address" cfg.ogmiosAddress ]
+        [ "--ogmios-port" cfg.ogmiosPort ]
+        [ "--db-connection" cfg.dbConnection ]
+        (optional cfg.useLatest "--use-latest")
+        (optional cfg.fromOrigin "--from-origin")
+        (optional cfg.fromTip "--from-tip")
+        (optionals (cfg.blockHash != null) [ "--block-hash" cfg.blockHash ])
+        (optionals (cfg.blockSlot != null) [ "--block-slot" cfg.blockSlot ])
```

Comment:
(`--from-origin`), (`--from-tip`), and (`--block-hash`, `--block-slot`) are exclusive, you can provide only one of these or app will error.

Reply:
https://github.com/mlabs-haskell/ogmios-datum-cache/pull/122/files/d45bf963d5b1af4c949cae37db875907c598bf41#diff-22b880528886b21f02995d2eb7b2baa3fda419007aa189fae7b76826329e27dcR118